File: StackOverflow/Q24016744/Q24016744.jl
# StackExchange Computational Science Q24016744
# https://stackoverflow.com/questions/24016744
# Creating Filter's Laplacian Matrix and Solving the Linear Equation for Image Filtering.
# References:
#   1.  A
# Remarks:
#   1.  Use in Julia as following:
#       -   Move to folder using `cd(raw"<PathToFolder>");`.
#       -   Activate the environment using `] activate .`.
#       -   Instantiate the environment using `] instantiate`.
#   2.  Using `Symmetric(BandedMatrix(mA));` for teh 5 point operator seems to be slow in Julia.  
#       Namely, it has no performance improvement over using the direct solver on `mA` itself.
#       It takes ~500 [Sec].
# TODO:
# 	1.  C
# Release Notes Royi Avital RoyiAvital@yahoo.com
# - 1.0.000     20/07/2024  Royi Avital
#   *   First release.

## Packages

# Internal
using LinearAlgebra;
using Printf;
using Random;
# External
using BandedMatrices;
using BenchmarkTools;
using ColorTypes;          #<! Required for Image Processing
using ConjugateGradients;
using FileIO;              #<! Required for loading images
using Krylov;
using LimitedLDLFactorizations;
using LinearOperators;
using LoopVectorization;   #<! Required for Image Processing
using MKLSparse;
using PlotlyJS;
using SparseArrays;
using StableRNGs;
using StaticKernels;       #<! Required for Image Processing


## Constants & Configuration
RNG_SEED = 1234;

juliaCodePath = joinpath(".", "..", "..", "JuliaCode");
include(joinpath(juliaCodePath, "JuliaInit.jl"));
include(joinpath(juliaCodePath, "JuliaImageProcessing.jl"));
include(joinpath(juliaCodePath, "JuliaVisualization.jl")); #<! Display Images

## General Parameters

figureIdx = 0;

exportFigures = true;

oRng = StableRNG(1234);

## Functions

struct DerivativeKernels{T <: AbstractFloat}
    mKx  :: Matrix{T}
    mK⁺x :: Matrix{T}
    mKy  :: Matrix{T}
    mK⁺y :: Matrix{T}
    DerivativeKernels{T}() where {T <: AbstractFloat} = new([one(T) -one(T)], [-one(T) one(T)], [one(T); -one(T);;], [-one(T); one(T);;]);
end

DerivativeKernels(dataType :: Type{<: AbstractFloat} = Float64) = DerivativeKernels{dataType}();

## Parameters

imgUrl = raw"https://i.imgur.com/69CLVcn.png" #<! "Flower.png" locally

λ = 0.05;
# Using `γ` instead of `γ` as `γ` reserved for the BLAS operation
γ = 0.95;
ε = 1e-5;

# Solver Parameters
numItr = 5_000;
solThr = 5e-7;


#%% Load / Generate Data

mI = load(download(imgUrl));
mI = ConvertJuliaImgArray(mI);
mI = mI ./ 255.0;

numRows = size(mI, 1);
numCols = size(mI, 2);

numPx = numRows * numCols;

sK = DerivativeKernels(Float64);

mO = similar(mI);
mTx = zeros(numRows, numCols - 1);
mTy = zeros(numRows - 1, numCols);
mDx = similar(mI);
mDy = similar(mI);
vY = zeros(numPx);
copyto!(vY, mI); #<! `copyto!()` ignores the shape

## Analysis
# The matrix: (C + D + λ * I) * x = y => A x = y

# Define the A operator

function OpAMul!( vY :: AbstractVector{T}, vX :: AbstractVector{T}, α :: T, β :: T, sK :: DerivativeKernels{T}, λ :: T, γ :: T, ε :: T, mO :: AbstractMatrix{T} ) where {T <: AbstractFloat}

    # Implements the Linear Operator (`mA`) by convolution operations.
    # Using `λ` instead of `α` as `α` reserved for the BLAS operation
    
    mX = @invoke reshape(vX, size(mO));

    mDx = Conv2D(mX, sK.mKx; convMode = CONV_MODE_VALID);
    # mAx = exp.(-(mDx .* mDx) ./ (2γ²));
    mAx = inv.((abs.(mDx) .^ γ) .+ ε)
    mDx .*= mAx;
    mDx = Conv2D(mDx, sK.mK⁺x; convMode = CONV_MODE_FULL);
    
    mDy = Conv2D(mX, sK.mKy; convMode = CONV_MODE_VALID);
    # mAy = exp.(-(mDy .* mDy) ./ (2γ²));
    mAy = inv.((abs.(mDy) .^ γ) .+ ε)
    mDy .*= mAy;
    mDy = Conv2D(mDy, sK.mK⁺y; convMode = CONV_MODE_FULL);

    mO .= mX .+ λ .* (mDx .+ mDy);
    if (β == zero(T))
        if (α == zero(T))
            vY .= zero(T);
        elseif (α == one(T))
            copyto!(vY, mO); #<! `copyto!()` ignores the shape
        else
            vY .= α .* mO[:];
        end
    elseif (β == one(T))
        if (α == one(T))
            vY .+= mO[:];
        else
            vY .+= α .* mO[:];
        end
    else
        if (α == zero(T))
            vY .*= β;
        elseif (α == one(T))
            vY .= (β .* vY) .+ mO[:];
        else
            vY .= (β .* vY) .+ α .* mO[:];
        end
    end

end

function OpAMul!!( vY :: AbstractVector{T}, vX :: AbstractVector{T}, α :: T, β :: T, sK :: DerivativeKernels{T}, λ :: T, γ :: T, ε :: T, mTx :: AbstractMatrix{T}, mTy :: AbstractMatrix{T}, mDx :: AbstractMatrix{T}, mDy :: AbstractMatrix{T}, mO :: AbstractMatrix{T} ) where {T <: AbstractFloat}

    # Non allocating version of `OpAMul!()` (Some allocations by `reshape()`)
    # Using `λ` instead of `α` as `α` reserved for the BLAS operation
    
    mX = @invoke reshape(vX, size(mO));

    mTx = Conv2D!(mTx, mX, sK.mKx; convMode = CONV_MODE_VALID);
    mTx ./= (abs.(mTx) .^ γ) .+ ε;
    mDx = Conv2D!(mDx, mTx, sK.mK⁺x; convMode = CONV_MODE_FULL);
    
    mTy = Conv2D!(mTy, mX, sK.mKy; convMode = CONV_MODE_VALID);
    mTy ./= (abs.(mTy) .^ γ) .+ ε;
    mDy = Conv2D!(mDy, mTy, sK.mK⁺y; convMode = CONV_MODE_FULL);

    mO .= mX .+ λ .* (mDx .+ mDy);
    if (β == zero(T))
        if (α == zero(T))
            vY .= zero(T);
        elseif (α == one(T))
            copyto!(vY, mO); #<! `copyto!()` ignores the shape
        else
            vY .= α .* mO[:];
        end
    elseif (β == one(T))
        if (α == one(T))
            vY .+= mO[:];
        else
            vY .+= α .* mO[:];
        end
    else
        if (α == zero(T))
            vY .*= β;
        elseif (α == one(T))
            vY .= (β .* vY) .+ mO[:];
        else
            vY .= (β .* vY) .+ α .* mO[:];
        end
    end

end

function GenMatA( vY :: AbstractVector{T}, numRows :: N, numCols :: N, λ :: T, γ :: T, ε :: T ) where {T <: AbstractFloat, N <: Integer}

    mDx = GenConvMtx([1.0 -1.0], numRows, numCols; convMode = CONV_MODE_VALID); #<! Horizontal Derivative 
    mDy = GenConvMtx([1.0; -1.0;;], numRows, numCols; convMode = CONV_MODE_VALID); #<! Vertical Derivative
    
    # mAx = Diagonal(exp.(- ((mDx * vY) .^ 2) ./ (2 * γ * γ)));
    # mAy = Diagonal(exp.(- ((mDy * vY) .^ 2) ./ (2 * γ * γ)));
    
    # Could apply λ in this step
    mAx = Diagonal(inv.((abs.(mDx * vY) .^ γ) .+ ε));
    mAy = Diagonal(inv.((abs.(mDy * vY) .^ γ) .+ ε));
    
    mA = (I + λ * (mDy' * mAy * mDy + mDx' * mAx * mDx));

    return mA;

end

function GenMatAxAy( vY :: AbstractVector{T}, numRows :: N, numCols :: N, λ :: T, γ :: T, ε :: T ) where {T <: AbstractFloat, N <: Integer}

    mDx = GenConvMtx([1.0 -1.0], numRows, numCols; convMode = CONV_MODE_VALID);
    mDy = GenConvMtx([1.0; -1.0;;], numRows, numCols; convMode = CONV_MODE_VALID);
    
    # mAx = Diagonal(exp.(- ((mDx * vY) .^ 2) ./ (2 * γ * γ)));
    # mAy = Diagonal(exp.(- ((mDy * vY) .^ 2) ./ (2 * γ * γ)));
    
    # Could apply λ in this step
    mAx = Diagonal(inv.((abs.(mDx * vY) .^ γ) .+ ε));
    mAy = Diagonal(inv.((abs.(mDy * vY) .^ γ) .+ ε));

    return mAx, mAy;

end


# %% Verify Operators
# Verify the convolution based operators and measure run time.

# Operator as a Matrix
mA = GenMatA(vY, numRows, numCols, λ, γ, ε); #<! Realization of the operator as a matrix

# Operator by Convolution
# No need for `MulAT!()` as the operator is symmetric
MulA!( vY :: AbstractVector{T}, vX :: AbstractVector{T}, α :: T, β :: T ) where {T <: AbstractFloat} = OpAMul!(vY, vX, α, β, sK, λ, γ, ε, mO);
MulA!( vY :: AbstractVector{T}, vX :: AbstractVector{T}, α :: T, β :: T ) where {T <: AbstractFloat} = OpAMul!!(vY, vX, α, β, sK, λ, γ, ε, mTx, mTy, mDx, mDy, mO);

# The operator is SPD (Symmetric Positive Definite)
isSymmetric = true;
isHermitian = false; #<! Irrelevant
opA = LinearOperator(Float64, numRows * numCols, numRows * numCols, isSymmetric, isHermitian, MulA!);

# Applying operators
vT1 = mA * vY;
vT2 = opA * vY;

opErr = maximum(abs.(vT1 - vT2));
resAnalysis = @sprintf("Maximum absolute deviation of operators: %0.5f", opErr);
println(resAnalysis);


# Timing Operators
runTime = @belapsed GenMatA(vY, numRows, numCols, λ, γ, ε) * vB setup = (vB = copy(vY)) seconds = 2;
resAnalysis = @sprintf("Matrix operator run time: %0.5f [Sec]", runTime);
println(resAnalysis);


runTime = @belapsed opA * vB setup = (vB = copy(vY)) seconds = 2;
resAnalysis = @sprintf("Functional operator run time: %0.5f [Sec]", runTime);
println(resAnalysis);


# Solutions Analysis

# Direct Solution
mA = GenMatA(vY, numRows, numCols, λ, γ, ε);
vXD = mA \ vY;
runTime = @belapsed GenMatA(vY, numRows, numCols, λ, γ, ε) \ vB setup = (vB = copy(vY)); seconds = 2;
resAnalysis = @sprintf("Direct solution run time: %0.5f [Sec]", runTime);
println(resAnalysis);

# Solution by `Krylov.jl`
oKrylovSolver = CgSolver(opA, vY);
vXKry = copy(vY);
Krylov.cg!(oKrylovSolver, opA, vY, vXKry);
runTime = @belapsed Krylov.cg!(oKrylovSolver, opA, vB, vXKry; atol = solThr, rtol = solThr, itmax = numItr) setup = begin (vB = copy(vY)); (vXKry = copy(vY)); end seconds = 2;
resAnalysis = @sprintf("Conjugate Gradient (`Krylov.jl`) solution run time: %0.5f [Sec]", runTime);
println(resAnalysis);

# Solution by `ConjugateGradients.jl`
hOpA( vY :: AbstractVector{T}, vX :: AbstractVector{T} ) where {T <: AbstractFloat} = OpAMul!!(vY, vX, 1.0, 0.0, sK, λ, γ, ε, mTx, mTy, mDx, mDy, mO);
oCG = CGData(numPx, eltype(vY));

vXCG = copy(vY);
# vX, exitCode, numItr = ConjugateGradients.cg(hOpA, vY; tol = solThr, maxIter = numItr, data = oCG);
exitCode, numItr = ConjugateGradients.cg!(hOpA, vY, vXCG; tol = solThr, maxIter = numItr, data = oCG);
runTime = @belapsed ConjugateGradients.cg!(hOpA, vY, vXCG; tol = solThr, maxIter = numItr, data = oCG) setup = (vXCG = copy(vY)) seconds = 2;
resAnalysis = @sprintf("Conjugate Gradient (`ConjugateGradients.jl`) solution run time: %0.5f [Sec]", runTime);
println(resAnalysis);


# Solution by Banded Matrix
# Seems to be much slower! Do not run this!
# mA = GenMatA(vY, numRows, numCols, λ, γ, ε);
# mB = Symmetric(BandedMatrix(mA));
# vXD = mB \ vY;
# runTime = @belapsed Symmetric(BandedMatrix(GenMatA(vY, numRows, numCols, λ, γ, ε))) \ vB setup = (vB = copy(vY)) seconds = 2;
# resAnalysis = @sprintf("Direct solution (`BandedMatrices.jl`) run time: %0.5f [Sec]", runTime);
# println(resAnalysis);

#  Preconditioner
mA = GenMatA(vY, numRows, numCols, λ, γ, ε); #<! Realization of the operator as a matrix
mP1 = lldl(mA);
mP1.D .= abs.(mP1.D);

mD  = Diagonal(mA);
mP2 = Diagonal(inv.(mD.diag));


## Display Results

figureIdx += 1;

hP = DisplayImage(mI; titleStr = "Input Image");
display(hP);

if (exportFigures)
    figFileNme = @sprintf("Figure%04d.png", figureIdx);
    savefig(hP, figFileNme);
end

figureIdx += 1;

copyto!(mO, vXD);
clamp!(mO, 0.0, 1.0);

hP = DisplayImage(mO; titleStr = "Output Image (Direct Solver), λ = $(λ)");
display(hP);

if (exportFigures)
    figFileNme = @sprintf("Figure%04d.png", figureIdx);
    savefig(hP, figFileNme);
end

figureIdx += 1;

copyto!(mO, oKrylovSolver.x);
clamp!(mO, 0.0, 1.0);

hP = DisplayImage(mO; titleStr = "Output Image (`Krylov.jl`), λ = $(λ)");
display(hP);

if (exportFigures)
    figFileNme = @sprintf("Figure%04d.png", figureIdx);
    savefig(hP, figFileNme);
end

figureIdx += 1;

copyto!(mO, vXCG);
clamp!(mO, 0.0, 1.0);

hP = DisplayImage(mO; titleStr = "Output Image (`ConjugateGradients.jl`), λ = $(λ)");
display(hP);

if (exportFigures)
    figFileNme = @sprintf("Figure%04d.png", figureIdx);
    savefig(hP, figFileNme);
end

File: StackOverflow/Q24016744/Q24016744.m
% StackOverflow Q24016744
% https://stackoverflow.com/questions/24016744
% Creating Filter's Laplacian Matrix and Solving the Linear Equation for Image Filtering
% References:
%   1.  
% Remarks:
%   1.  B
% TODO:
% 	1.  C
% Release Notes Royi Avital RoyiAvital@yahoo.com
% - 1.0.000     25/08/2024
%   *   First release.


%% General Parameters

subStreamNumberDefault = 0;

run('InitScript.m');

figureIdx           = 0;
figureCounterSpec   = '%04d';

generateFigures = ON;

%% Constants



%% Parameters

% Data
% From Image Smoothing via L0 Gradient Minimization (https://www.cse.cuhk.edu.hk/~leojia/projects/L0smoothing/ImageSmoothing.htm)
imgUrl = 'https://i.imgur.com/8LuDAju.png'; %<! Basketball
% imgUrl = 'https://i.imgur.com/LzBNqV8.png'; %<! Beach
% imgUrl = 'https://i.imgur.com/69CLVcn.png'; %<! Flower
% imgUrl = 'https://i.imgur.com/r9nUFsM.png'; %<! Mountain

% Model
paramLambda = 2.15;
paramGamma  = 1.5;
paramEps    = 1e-4;

% Solver
numIter = 5000;
solThr  = 1e-6;


%% Generate / Load Data

mI = im2double(imread(imgUrl));
mI = mean(mI, 3);

numRows = size(mI, 1);
numCols = size(mI, 2);

numPx = numRows * numCols;

% mI = rand(numRows, numCols);

vY = mI(:);


%% Analysis

mA = GenMatA(vY, numRows, numCols, paramLambda, paramGamma, paramEps);

[mAx, mAy] = GenMatAxAy(vY, numRows, numCols, paramGamma, paramEps);


hOpA = @(vT) ApplyOpA(vT, numRows, numCols, mAx, mAy, paramLambda);

% Applying operators
vT1 = mA * vY;
vT2 = hOpA(vY);

max(abs(vT1 - vT2), [], 'all')

for ii = 1:2
    vT1 = mA * vT1;
    vT2 = hOpA(vT2);
end

max(abs(vT1 - vT2), [], 'all')

% hF = @() mA * vY;
hF = @() GenMatA(vY, numRows, numCols, paramLambda, paramGamma, paramEps) * vY;
hG = @() hOpA(vY);

TimeItMin(hF, 1)
TimeItMin(hG, 1)

condest(mA)

% Direct Solution
vXD = mA \ vY;
mL = ichol(mA, struct('michol', 'on'));
vX = pcg(hOpA, vY, solThr, numIter);
% vX = pcg(hOpA, vY, solThr, numIter, mL, mL.');
% vX = pcg(mA, vY, solThr, numIter);
% vX = pcg(mA, vY, solThr, numIter, mL, mL.');

hF = @() mA \ vY;
hG = @() pcg(hOpA, vY, solThr, numIter);

TimeItMin(hF, 1)
TimeItMin(hG, 1)


figure();
imshow(reshape(vY, numRows, numCols), [0, 1]);

figure();
imshow(reshape(vXD, numRows, numCols), [0, 1]);

figure();
imshow(reshape(vX, numRows, numCols), [0, 1]);

sqrt(mean(abs(vX - vXD) .^ 2))



%% Display Results

% figureIdx = figureIdx + 1;
% 
% hFigure = figure('Position', figPosLarge);
% hAxes   = axes(hFigure);
% set(hAxes, 'NextPlot', 'add');
% hLineObj = plot(mX);
% for ii = 1:numSignals
%     set(hLineObj(ii), 'DisplayName', ['Line ', num2str(ii)]);
% end
% set(hLineObj, 'LineWidth', lineWidthNormal);
% hLineObj = plot(vX, 'DisplayName', 'Mean Line');
% set(hLineObj, 'LineWidth', lineWidthNormal, 'LineStyle', ':');
% 
% set(get(hAxes, 'Title'), 'String', {['Estimate Data by Partial Data']}, ...
%     'FontSize', fontSizeTitle);
% set(get(hAxes, 'XLabel'), 'String', {['Sample Index']}, ...
%     'FontSize', fontSizeAxis);
% set(get(hAxes, 'YLabel'), 'String', {['Sample Value']}, ...
%     'FontSize', fontSizeAxis);
% 
% hLegend = ClickableLegend();
% 
% if(generateFigures == ON)
%     % saveas(hFigure, ['Figure', num2str(figureIdx, figureCounterSpec), '.png']);
%     print(hFigure, ['Figure', num2str(figureIdx, figureCounterSpec), '.png'], '-dpng', '-r0'); %<! Saves as Screen Resolution
% end
% 
% 
% figureIdx = figureIdx + 1;
% 
% hFigure = figure('Position', figPosLarge);
% hAxes   = axes(hFigure);
% set(hAxes, 'NextPlot', 'add');
% hLineObj = plot(mX);
% for ii = 1:numSignals
%     set(hLineObj(ii), 'DisplayName', ['Line ', num2str(ii)]);
% end
% set(hLineObj, 'LineWidth', lineWidthNormal);
% hLineObj = plot(vXX, 'DisplayName', 'Optimized Line');
% set(hLineObj, 'LineWidth', lineWidthNormal, 'LineStyle', ':');
% 
% set(get(hAxes, 'Title'), 'String', {['Estimate Data by Partial Data']}, ...
%     'FontSize', fontSizeTitle);
% set(get(hAxes, 'XLabel'), 'String', {['Sample Index']}, ...
%     'FontSize', fontSizeAxis);
% set(get(hAxes, 'YLabel'), 'String', {['Sample Value']}, ...
%     'FontSize', fontSizeAxis);
% 
% hLegend = ClickableLegend();
% 
% if(generateFigures == ON)
%     % saveas(hFigure, ['Figure', num2str(figureIdx, figureCounterSpec), '.png']);
%     print(hFigure, ['Figure', num2str(figureIdx, figureCounterSpec), '.png'], '-dpng', '-r0'); %<! Saves as Screen Resolution
% end


%% Auxiliary Functions

function [ mA ] = GenMatA( vY, numRows, numCols, paramLambda, paramGamma, paramEps )

numElements = numRows * numCols;

mDx = CreateConvMtx2D([1, -1], numRows, numCols, 3); %<! 3 -> Conv Shape Valid
mDy = CreateConvMtx2D([1; -1], numRows, numCols, 3); %<! 3 -> Conv Shape Valid

% Could apply λ in this step
mAx = spdiags(1 ./ ((abs(mDx * vY) .^ paramGamma) + paramEps), 0, numElements - numRows, numElements - numRows);
mAy = spdiags(1 ./ ((abs(mDy * vY) .^ paramGamma) + paramEps), 0, numElements - numCols, numElements - numCols);
    
mA = speye(numElements) + (paramLambda * (mDy.' * mAy * mDy + mDx.' * mAx * mDx));

end

function [ mAx, mAy ] = GenMatAxAy( vY, numRows, numCols, paramGamma, paramEps )

mY = reshape(vY, numRows, numCols);
mIx = conv2(mY, [1, -1], 'valid');
mIy = conv2(mY, [1; -1], 'valid');

mAx = 1 ./ ((abs(mIx) .^ paramGamma) + paramEps);
mAy = 1 ./ ((abs(mIy) .^ paramGamma) + paramEps);


end


function [ vX ] = ApplyOpA( vY, numRows, numCols, mAx, mAy, paramLambda )

mY = reshape(vY, numRows, numCols);
mIx = conv2(mY, [1, -1], 'valid');
mIy = conv2(mY, [1; -1], 'valid');

% Should be calculated once based on the original `vY`.
% In PCG each iteration `vY` changes, so it must be calculated as an
% auxiliary function.
% mIx = mIx ./ ((abs(mIx) .^ paramGamma) + paramEps);
% mIy = mIy ./ ((abs(mIy) .^ paramGamma) + paramEps);

mIx = mIx .* mAx;
mIy = mIy .* mAy;

mIx = conv2(mIx, [-1, 1], 'full');
mIy = conv2(mIy, [-1; 1], 'full');

vX = vY + (paramLambda * (mIx(:) + mIy(:)));


end


%% Restore Defaults

% set(0, 'DefaultFigureWindowStyle', 'normal');
% set(0, 'DefaultAxesLooseInset', defaultLoosInset);

